File: python/taco/reader/dataset.py
from __future__ import annotations

from ..contract.collection import Collection
from ..contract.contract import Contract
from ..errors import CollectionError, ContainerError
from .collection import merge_collections
from .manifest import resolve_dataset
from .source import Source


class Dataset:
    def __init__(self, source: Source) -> None:
        resolution = resolve_dataset(source)
        self.sources = resolution.sources
        if resolution.collection is None:
            self.collection = merge_collections(self.sources)
        else:
            try:
                self.collection = Collection.from_dict(resolution.collection)
            except CollectionError as error:
                raise ContainerError(f"version {resolution.version!r} embeds an invalid collection: {error}") from error
        self.version = resolution.version or self.collection.dataset_version
        self.versions = resolution.versions
        self.manifest = resolution.manifest

    @property
    def contract(self) -> Contract:
        return self.collection.contract

    def __repr__(self) -> str:
        location = f"source={self.sources[0]!r}" if len(self.sources) == 1 else f"sources={len(self.sources)}"
        return f"Dataset({self.collection.id!r}, version={self.collection.dataset_version!r}, {location})"

    def _repr_html_(self) -> str:
        from .._repr import dataset_html

        return dataset_html(self)


__all__ = ["Dataset"]

File: julia/src/reader/dataset.jl
struct Contract
    structure::Union{Nothing,Vector{String}}
    metadata::Dict{String,Any}
    derived::Dict{String,Any}
    levels::Vector{String}
end


struct Dataset
    sources::Vector{String}
    collection::Dict{String,Any}
    contract::Contract
    version::String
    versions::Vector{String}
    manifest::Union{Nothing,String}
end


Dataset(sources, collection, contract) = Dataset(
    sources,
    collection,
    contract,
    String(collection["dataset_version"]),
    String[],
    nothing,
)


function _build_contract(collection, levels)
    raw_structure = collection["taco:structure"]
    structure = raw_structure === nothing ? nothing : String.(raw_structure)
    metadata = Dict{String,Any}(collection["taco:metadata"])
    derived = Dict{String,Any}(get(collection, "taco:derived", Dict{String,Any}()))
    return Contract(structure, metadata, derived, levels)
end


function _open_dataset(source)
    resolution = _resolve_dataset(source)
    sources = resolution.sources
    collection = resolution.collection
    levels = String[]
    if collection === nothing
        parsed = _parse_collection.(_collection_documents(sources))
        collection = _merge_collections(first.(parsed), last.(parsed), sources)
        levels = last(first(parsed))
    else
        levels = resolution.levels
    end
    version = something(resolution.version, String(collection["dataset_version"]))
    return Dataset(
        sources,
        collection,
        _build_contract(collection, levels),
        version,
        resolution.versions,
        resolution.manifest,
    )
end


function Base.show(io::IO, dataset::Dataset)
    collection = dataset.collection
    print(io, "Taco.Dataset(")
    show(io, collection["id"])
    print(
        io,
        ", version=",
        collection["dataset_version"],
        ", sources=",
        length(dataset.sources),
        ")",
    )
end
